### PR TITLE
Atomic index rebuilding for Elasticsearch

### PR DESCRIPTION
**Pull request by kaedroho**
This pull request implements an "atomic index rebuilder" that can be enabled on a per-index basis with the "ATOMIC_REBUILD" option in ``WAGTAILSEARCH_BACKENDS``.

The "basic" index rebuilder (current behaviour) deletes the index and reloads everything again from scratch. This leads to users having to use a half-populated index during the rebuild and can also cause problems like #1320.

The new "atomic" index rebuilder creates a separate index and loads everything into that. The old index will continue to be used until the new one is fully rebuilt. When the rebuild is complete, the indices are atomically swapped then the old index is deleted. This is implemented using Elasticsearch aliases.

https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html

Index aliases are supported by searchly (and of course, self-hosted instances too) so this shouldn't be an issue for most users.

This feature is switched off by default and should be considered "experimental" for a couple of releases so we can make sure any issues are properly ironed out.

### CODE REVIEW

**Comment by kaedroho** [comment]
Ready for review!

File: wagtail/wagtailsearch/backends/elasticsearch.py [hunk 1]
```diff
@@ -307,6 +354,115 @@ def _do_count(self):
...
+
+        # Make sure there isn't currently an index that clashes with alias_name
+        # This can happen when the atomic rebuilder is first enabled
+        if not self.es.indices.exists_alias(self.alias_name):
+            try:
+                self.es.indices.delete(self.alias_name)
+
+                # An index was deleted so there couldn't have been an alias there.
+                # Create an alias now so the site search doesn't break
+                self.es.indices.put_alias(name=self.alias_name, index=self.index_name)
```
Comment (gasman) [inline wagtail/wagtailsearch/backends/elasticsearch.py]: This confused me for a while - I wasn't sure why we were creating the alias at this point (and only when the deletion of a same-named index has succeeded). I think it was the comment that threw me - I'm guessing that was written before the `exists_alias` check was added...

Is there any reason why we can't move this whole block to the `finish` method? That way, we don't need this extra `put_alias`, and it means that the site search can continue to work (with a complete index) during the first rebuild after enabling ATOMIC_REBUILD.

File: wagtail/wagtailsearch/backends/elasticsearch.py [hunk 2]
```diff
@@ -307,6 +354,115 @@ def _do_count(self):
...
+        self.es.indices.refresh(self.index_name)
+
+        # Find index that alias currently points to
+        old_index = set(self.es.indices.get_alias(name=self.alias_name).keys()) - {self.index_name}
+
+        # Update alias to point to new index
+        self.es.indices.put_alias(name=self.alias_name, index=self.index_name)
+
+        # Delete old index
+        # es.indicies.get_alias can return multiple indicies. Delete them all
```
Comment (gasman) [inline wagtail/wagtailsearch/backends/elasticsearch.py]: 'indicies' should be 'indices'

**Comment by gasman** [comment]
Perhaps this would be a good time to update the sentence "The search may not return any results while this command is running, so avoid running it at peak times." on http://docs.wagtail.io/en/v1.0/topics/search/indexing.html#the-update-index-command - it probably just needs qualifying with something like: "With certain backends and configurations, the search may not return any results...". (Or is it better to wait until this feature is declared non-experimental?)

**Comment by gasman** [comment]
...other than the above, all looks good to me!

File: wagtail/wagtailsearch/backends/elasticsearch.py (hunk 1 above)
Comment (kaedroho) [inline wagtail/wagtailsearch/backends/elasticsearch.py]: Apologies for that. Yep, I'll move it to the finish method

**Comment by kaedroho** [comment]
Above issues fixed.

I expect that 90% of people would still use the old behaviour until we make atomic rebuild the default. So I think we should leave that message as it is for now.

**Pull request merged**